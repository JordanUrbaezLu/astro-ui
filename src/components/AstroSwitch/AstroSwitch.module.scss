@use "../../utility/globals";

.astroSwitchPill {
  background: none;
  border: 0;
  cursor: pointer;
  display: flex;
  font-family: Verdana, Geneva, Tahoma, sans-serif;
  margin: 0;
  padding: 0;
  align-items: center;
  font-size: 16px;
}

.astroSwitchIndicator {
  background: gray;
  border-radius: 999px;
  position: relative;
  height: 22px;
  width: 42px;
  transition: all 0.15s ease-in;

  &.primary {
    &::before {
      background: white;
      border-radius: 999px;
      content: "";
      display: inline-block;
      top: 3px;
      right: 23px;
      height: 16px;
      width: 16px;
      position: absolute;
      transition: all 0.15s ease-in;
    }
  }

  &.secondary {
    &::before {
      background: white;
      border-radius: 999px;
      content: "";
      display: inline-block;
      top: 3px;
      right: 23px;
      height: 16px;
      width: 16px;
      position: absolute;
      transition: all 0.15s ease-in;
    }
  }

  &.on {
    &.primary {
      background: globals.$primary-color;
    }

    &.secondary {
      background: globals.$secondary-color;
    }

    &::before {
      transform: translateX(20px);
    }
  }
}

.astroSwitchLabel {
  color: gray;
  margin-left: 8px;
  transition: all 0.15s ease-in;

  &.on {
    &.primary {
      color: globals.$primary-color;
    }

    &.secondary {
      color: globals.$secondary-color;
    }
  }
}
