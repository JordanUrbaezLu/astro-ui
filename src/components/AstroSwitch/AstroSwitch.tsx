import classNames from "classnames";
import * as React from "react";
import styles from "./AstroSwitch.module.scss";

export type AstroSwitchVariant = "primary" | "secondary";

export interface AstroSwitchProps {
  /**
   * If the AstroSwitch is on
   *
   * @default false
   */
  isOn?: boolean;
  /**
   * The label for the AstroSwitch
   */
  label?: string;
  /**
   * The callback fired when the AstroSwitch is clicked
   */
  onClick: () => void;
  /**
   * The variant of the AstroSwitch
   *
   * @default "primary"
   */
  variant?: AstroSwitchVariant;
}

export const AstroSwitch: React.FC<AstroSwitchProps> = (props) => {
  const { isOn = false, label, onClick, variant = "primary", ...rest } = props;

  const astroSwitchPill = classNames(
    styles.astroSwitchPill,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    isOn && styles.on
  );

  const astroSwitchIndicator = classNames(
    styles.astroSwitchIndicator,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    isOn && styles.on
  );

  const astroSwitchLabel = classNames(
    styles.astroSwitchLabel,
    variant === "primary" && styles.primary,
    variant === "secondary" && styles.secondary,
    isOn && styles.on
  );

  return (
    <button
      aria-checked={isOn}
      className={astroSwitchPill}
      onClick={onClick}
      role="switch"
      type="button"
      {...rest}
    >
      <span className={astroSwitchIndicator} />
      {label && <span className={astroSwitchLabel}>{label}</span>}
    </button>
  );
};
