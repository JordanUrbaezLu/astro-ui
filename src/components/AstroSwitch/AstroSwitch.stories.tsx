import { Story } from "@storybook/react";
import * as React from "react";
import { AstroSwitch, AstroSwitchProps } from "./AstroSwitch";

export default {
  component: AstroSwitch,
  title: "Components/AstroSwitch",
};

const Template: Story<Omit<AstroSwitchProps, "isOn" | "onClick">> = (args) => {
  const [astroSwitchState, toggleAstroSwitchState] =
    React.useState<boolean>(true);

  return (
    <AstroSwitch
      {...args}
      isOn={astroSwitchState}
      onClick={() => toggleAstroSwitchState(!astroSwitchState)}
    />
  );
};

export const Default = Template.bind({});

export const Primary = Template.bind({});
Primary.args = {
  variant: "primary",
};

export const Secondary = Template.bind({});
Secondary.args = {
  variant: "secondary",
};

export const PrimaryAndLabel = Template.bind({});
PrimaryAndLabel.args = {
  label: "Astro Switch Label",
  variant: "primary",
};

export const SecondaryAndLabel = Template.bind({});
SecondaryAndLabel.args = {
  label: "Astro Switch Label",
  variant: "secondary",
};
