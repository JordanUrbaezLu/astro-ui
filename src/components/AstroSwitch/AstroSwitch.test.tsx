import { render, screen } from "@testing-library/react";
import { AstroSwitch, AstroSwitchVariant } from "./AstroSwitch";
import { axe } from "jest-axe";

describe("AstroSwitch", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <AstroSwitch isOn onClick={jest.fn()} label="Astro Switch Label" />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <AstroSwitch isOn onClick={jest.fn()} label="Astro Switch Label" />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe("Props", () => {
  test("Should render label correctly.", () => {
    const label = "Astro Switch Label";

    render(<AstroSwitch isOn onClick={jest.fn()} label={label} />);
    expect(screen.getByRole("switch")).toHaveTextContent(label);
  });

  test.each<AstroSwitchVariant>(["primary", "secondary"])(
    "Should render variant correctly.",
    (variant) => {
      render(
        <AstroSwitch
          isOn
          onClick={jest.fn()}
          label="Astro Switch Label"
          variant={variant}
        />
      );
      expect(screen.getByRole("switch")).toHaveClass(variant);
    }
  );
});
