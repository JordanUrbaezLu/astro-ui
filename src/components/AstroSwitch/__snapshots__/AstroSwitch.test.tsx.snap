// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstroSwitch Should render component correctly. 1`] = `
<button
  aria-checked="true"
  class="astroSwitchPill primary on"
  role="switch"
  type="button"
>
  <span
    class="astroSwitchIndicator primary on"
  />
  <span
    class="astroSwitchLabel primary on"
  >
    Astro Switch Label
  </span>
</button>
`;
